dask/dataframe/core.py

```diff
@@ -4672,15 +4673,82 @@ def set_index(
         >>> # ddf2 = ddf.set_index("name", divisions=["Alice", "Laura", "Ursula", "Zelda"])
         """
+
+        # Deal with bool params
         if inplace:
             raise NotImplementedError("The inplace= keyword is not supported")
         pre_sorted = sorted
         del sorted
 
-        if isinstance(other, Series) and other._name == self.index._name:
-            # Index is equal to current index, no-op
-            return self
+        # Check other can be translated to column name or column object, possibly flattening it
+        if not isinstance(other, str):
+            if isinstance(other, Sequence):
+                # Accept ["a"], but not [["a"]]
+                if len(other) == 1 and (
+                    isinstance(other[0], str) or not isinstance(other[0], Sequence)
+                ):
+                    other = other[0]
+                else:
+                    raise NotImplementedError(
+                        "Dask dataframe does not yet support multi-indexes.\n"
+                        f"You tried to index with this index: {other}\n"
+                        "Indexes must be single columns only."
+                    )
+
+            # Can also be a frame
+            elif isinstance(other, DataFrame):
+                raise NotImplementedError(
+                    "Dask dataframe does not yet support multi-indexes.\n"
+                    f"You tried to index with a frame with these columns: {list(other.columns)}\n"
+                    "Indexes must be single columns only."
+                )
 
+        # If already a series
+        index, columns = self.index, self.columns
+        if is_series := isinstance(other, Series):
+            # If it's already the index, there's nothing to do
+            if other._name == index._name:
+                return self
+            # Otherwise, check length matches when other isn't one of the data columns
+            s = other
+            if not any(
+                s._name == col_series._name for _, col_series in self.items()
+            ) and (  # for cases with other=df.column
+                len_index := len(s)
+            ) != (
+                len_self := len(self)
+            ):
+                raise ValueError(
+                    f"Length mismatch: series to become index has {len_index} rows, but data have {len_self} rows"
+                )
+
+        # If the name of a column/index
+        else:
+            # If the index, no additional checks
+            if (
+                other == index.name
+            ):  # here it's name that should be checked, instead of _name
+                return self
+            # If a missing column, KeyError
+            if other not in columns:
+                raise KeyError(
+                    f"Data has no column '{other}': use either index '{index.name}' or any column of {list(columns)}"
+                )
+            s = self[other]
+
+        # Ensure there are no nulls
```

Comment:
This comment seems to be a replica of the code on the next line. Since the error message in a few lines is pretty clear about why we're doing this, can you remove the comment?

Reply:
Removed